fenicsadapter/fenicsadapter.py

```diff
@@ -1,759 +1,458 @@
+        my_u = user_u.copy()
+        assert (my_u != user_u)  # wrt to pointer, make sure the direct function reference is not stored
+        self._checkpoint = SolverState(my_u, t, n)
+        self._interface.mark_action_fulfilled(self.action_write_checkpoint())
 
-        # Avoid application of PointSource and Dirichlet boundary condition at the same point by filtering
-        x_forces = filter_point_sources(x_forces, self._Dirichlet_Boundary)
-        y_forces = filter_point_sources(y_forces, self._Dirichlet_Boundary)
-
-        return x_forces.values(), y_forces.values()  # don't return dictionary, but list of PointSources
-
-    def _restore_solver_state_from_checkpoint(self, state):
-        """Resets the solver's state to the checkpoint's state.
-        :param state: current state of the FEniCS solver
+    def retrieve_checkpoint(self):
         """
-        logger.debug("Restore solver state")
-        state.update(self._checkpoint.get_state())
-        self._interface.mark_action_fulfilled(precice.action_read_iteration_checkpoint())
-
-    def _advance_solver_state(self, state, u_np1, dt):
-        """Advances the solver's state by one timestep.
-        :param state: old state
-        :param u_np1: new value
-        :param dt: timestep size
-        :return:
+        Resets the FEniCS participant state to the state of the stored checkpoint.
+
+        Returns
+        -------
+        u : FEniCS Function
+            Current state of the physical variable of interest for this participant.
+        t : double
+            Current simulation time.
+        n : int
+            Current time window (iteration) number.
         """
-        logger.debug("Advance solver state")
-        logger.debug("old state: t={time}".format(time=state.t))
-        state.update(SolverState(u_np1, state.t + dt, state.n + 1))
-        logger.debug("new state: t={time}".format(time=state.t))
+        assert (not self.is_time_window_complete())  # avoids invalid control flow
+        logger.debug("Restore solver state")
+        self._interface.mark_action_fulfilled(self.action_read_checkpoint())
+        return self._checkpoint.get_state()
 
-    def _save_solver_state_to_checkpoint(self, state):
-        """Writes given solver state to checkpoint.
-        :param state: state being saved as checkpoint
+    def advance(self, dt):
         """
-        logger.debug("Save solver state")
-        self._checkpoint.write(state)
-        self._interface.mark_action_fulfilled(precice.action_write_iteration_checkpoint())
+        Advances coupling in preCICE.
 
-    def advance(self, write_function, u_np1, u_n, t, dt, n):
-        """Calls preCICE advance function using precice and manages checkpointing.
-        The solution u_n is updated by this function via call-by-reference. The corresponding values for t and n are returned.
+        Parameters
+        ----------
+        dt : double
+            Time step value used for the last iteration.
```

Comment:
```suggestion
    Length of timestep used by the solver.
```

Stolen from [`SolverInterface::advance(...)`](https://github.com/precice/precice/blob/657f7596b1428d5b26c97eeaa9c26c13b4a32d9e/src/precice/SolverInterface.hpp#L124-L143). Might also be worth considering to use the term `computedTimestepLength` for sake of consistency.